blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+        let api = self.api.read().await;
+        let actor = api.state_get_actor(&addr, cur_ts)?;
+        let mut base_sequence = actor.sequence;
+        // TODO here lotus has a todo, so I guess we should eventually remove cur_ts from one
+        // of the params for this method and just use the chain head
+        let msgs = api.messages_for_tipset(cur_ts)?;
+        drop(api);
+
+        for m in msgs {
+            if m.from() == addr {
+                if m.sequence() != base_sequence {
+                    return Err(Error::Other("tipset has bad sequence ordering".to_string()));
+                }
+                base_sequence += 1;
+            }
+        }
+
+        Ok(base_sequence)
+    }
+
+    /// Get the state balance for the actor that corresponds to the supplied address and tipset,
+    /// if this actor does not exist, return an error
+    async fn get_state_balance(&mut self, addr: &Address, ts: &Tipset) -> Result<BigInt, Error> {
+        let api = self.api.read().await;
+        let actor = api.state_get_actor(&addr, &ts)?;
+        Ok(BigInt::from(actor.balance))
+    }
+
+    /// Remove a message given a sequence and address from the messagepool
+    pub async fn remove(&mut self, from: &Address, sequence: u64) -> Result<(), Error> {
+        remove(from, self.pending.clone(), sequence).await
+    }
+
+    /// Return a tuple that contains a vector of all signed messages and the current tipset for
+    /// self.
+    pub async fn pending(&self) -> Result<(Vec<SignedMessage>, Tipset), Error> {
+        let mut out: Vec<SignedMessage> = Vec::new();
+
+        let pending = self.pending.read().await;
+        let pending_hm = pending.clone();
+        drop(pending);
+
+        for (addr, _) in pending_hm {
+            out.append(
+                self.pending_for(&addr)
+                    .await
+                    .ok_or_else(|| Error::InvalidFromAddr)?
+                    .as_mut(),
+            )
+        }
+
+        let cur_t = self.cur_tipset.read().await;
+        let cur_ts = cur_t.clone();
+        drop(cur_t);
+
+        Ok((out, cur_ts))
+    }
+
+    /// Return a Vector of signed messages for a given from address. This vector will be sorted by
+    /// each messsage's sequence. If no corresponding messages found, return None result type
+    async fn pending_for(&self, a: &Address) -> Option<Vec<SignedMessage>> {
+        let pending = self.pending.read().await;
+
+        let mset = pending.get(a);
+        match mset {
+            Some(msgset) => {
+                if msgset.msgs.is_empty() {
+                    return None;
+                }
+
+                let mut msg_vec = Vec::new();
+
+                for (_, item) in msgset.msgs.clone() {
+                    msg_vec.push(item);
+                }
+
+                msg_vec.sort_by_key(|value| value.message().sequence());
+
+                Some(msg_vec)
+            }
+            None => None,
+        }
+    }
```

Comment:
```suggestion
    async fn pending_for(&self, a: &Address) -> Option<Vec<SignedMessage>> {
        let pending = self.pending.read().await;
        let mset = pending.get(a)?;

        if mset.msgs.is_empty() {
            return None;
        }

        let mut msg_vec = Vec::new();

        for (_, item) in mset.msgs.iter() {
            msg_vec.push(item.clone());
        }

        msg_vec.sort_by_key(|value| value.message().sequence());

        Some(msg_vec)
    }
```

Few small changes so I'll just do as one. There was the cosmetic change to avoid match but also the functional one which is to avoid cloning the hashset instead of cloning the elements being pushed